config/coordinator.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"crypto/md5"
+	"encoding/binary"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// Coordinator coordinates Alertmanager configurations beyond the lifetime of a
+// single configuration.
+// TODO: Make Coordinator thread safe
+type Coordinator struct {
+	// TODO: Either do all cfg or all conf
+	cfgFilePath string
+	logger      log.Logger
+
+	config *Config
+
+	configHashMetric        prometheus.Gauge
+	configSuccessMetric     prometheus.Gauge
+	configSuccessTimeMetric prometheus.Gauge
+
+	subscribers []func(*Config)
+}
+
+// NewCoordinator returns a new coordinator with the given configuration file
+// path.
+func NewCoordinator(cfgFilePath string, r prometheus.Registerer, l log.Logger) *Coordinator {
+	c := &Coordinator{
+		cfgFilePath: cfgFilePath,
+		logger:      l,
+	}
+
+	c.registerMetrics(r)
+
+	return c
+}
+
+func (c *Coordinator) registerMetrics(r prometheus.Registerer) {
+	configHash := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_hash",
+		Help: "Hash of the currently loaded alertmanager configuration.",
+	})
+	configSuccess := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_last_reload_successful",
+		Help: "Whether the last configuration reload attempt was successful.",
+	})
+	configSuccessTime := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_last_reload_success_timestamp_seconds",
+		Help: "Timestamp of the last successful configuration reload.",
+	})
+
+	r.MustRegister(configHash, configSuccess, configSuccessTime)
+
+	c.configHashMetric = configHash
+	c.configSuccessMetric = configSuccess
+	c.configSuccessTimeMetric = configSuccessTime
+}
+
+// Subscribe subscribes the given Subscribers to configuration changes.
+func (c *Coordinator) Subscribe(ss ...func(*Config)) {
+	c.subscribers = append(c.subscribers, ss...)
+}
+
+func (c *Coordinator) notifySubscribers() {
+	for _, s := range c.subscribers {
+		s(c.config)
+	}
+}
+
+// loadFromFile triggers a configuration load, discarding the old configuration.
+func (c *Coordinator) loadFromFile() error {
+	level.Info(c.logger).Log(
+		"msg", "Loading configuration file",
+		"file", c.cfgFilePath,
+	)
+
+	conf, plainCFG, err := LoadFile(c.cfgFilePath)
```

Comment:
(nit) s/plainCFG/plainCfg/

Reply:
In general, Go would use `plainCFG` see: https://github.com/golang/go/wiki/CodeReviewComments#initialisms.

I decided to replace all `cfg` with `config` to stay consistent with `config.go`, let me know is you think differently.

Reply:
Hmm, I wouldn't consider `CFG` as an initialism or acronym but I'm not a linguistic expert :-)